### PR TITLE
BUG: Provide users with the ability to read TIFF images that have the tags ``XResolution`` or ``YResolution`` with 0 in the denominator.

### PR DESCRIPTION
**Pull request by hmaarrfk**
It seems that certain tiffs can have the tag but the denominator can be 0.

Not sure how to handle them correctly, but I'm leaving this here as a placeholder.

We can try to recreate the TIFF and add it to a test.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
It may be that a warning would be appropriate in this case.

Currently, it is causing some of images to simply be unable to opened with imageio

**Comment by FirefoxMetzger** [comment]
Thanks for the PR @hmaarrfk !

> It seems that certain tiffs can have the tag but the denominator can be 0.

Oha, I didn't know this was actually possible. At first sight, this looks like a corrupt/broken TIFF writer to me (resolution has to be a rational number) and my first stop would be to look for a bug in the code that produces the offending TIFF.

> Not sure how to handle them correctly, but I'm leaving this here as a placeholder.

I kind of like the idea of throwing an exception, at least by default. Anything else can easily cause invalid results that are hard to spot. 

Say you have a function that quantifies the size of some microbe in an image. If the TIFF specifies a resolution, it uses that to compute the size, if not, it uses a default value (1 along each dimension doesn't seem that unreasonable, right?). Now any scenario where ImageIO returns something but ends up ignoring the resolution metadata becomes dangerous because this function will happily use a (potentially very wrong) default value and report a result. We could emit a warning along the lines of "Warning: ignoring corrupt resolution metadata", but I know more than enough people (especially in research) that are happy to ignore runtime warnings as long as it still produces some number [it can't be THAT wrong, right?]. Because of that, I am a bit wary of not raising by default.

That said, if explicitly told to be lax I am happy to have the reader ignore corrupt metadata fields (or warn about them) because it does increase compatibility.

@hmaarrfk What do you think of adding a `metadata_policy="raise"` kwarg that takes the values `["raise", "warn", "ignore"]` to control the desired behavior. The default would remain `"strict"`, but those who need it (and dare) may choose weaker policies to tackle problematic files. If this works well and is useful for tifffile, we can also think of adopting this idea in other plugins. Also, naming suggestions and alternative ideas are welcome :)

> We can try to recreate the TIFF and add it to a test.

If you have an offending file at hand, that would make this easier indeed. I tried to reproduce this failure on master, but tifffile doesn't allow you to write div/0 as resolution metadata (nor does ImageIO):

```python
import tifffile
import imageio as iio
import numpy as np

img = np.arange(100).reshape(10,10)
tifffile.imwrite("corrupt.tiff", img, resolution=((10, 0),1))  # raises ZeroDivisionError
```

> It may be that a warning would be appropriate in this case.

See my earlier discussion. I think `raise` is the better default strategy, but I am open to warn for/ignore broken fields if the user explicitly asks for it.

> Currently, it is causing some of images to simply be unable to opened with imageio

For the v2 API I can understand this, because it has to load metadata for each image to populate `img.meta`. However, the v3 API should be unaffected, i.e.,

```python

import imageio.v3 as iio

img = iio.imread("offending.tiff")
```

should still work, because it doesn't touch metadata.

**Comment by hmaarrfk** [comment]
tbh, we found our offending "invalid" tiff writer. And we "fixed it internally". We hope we don't have invalid tiffs in the wild.

I'll give your proposals a read tomorrow.

**Comment by FirefoxMetzger** [comment]
I'm curious; since you put "invalid" in apostrophes, does it actually make sense to use a div/0 for your use-case?

**Comment by hmaarrfk** [comment]
No. We just wanted to make sure we didn't confuse it with the default of 150 dpi.

**Comment by hmaarrfk** [comment]
I think a simple test is missing, where we create a tiff, then try to load it. Are you able to contribute that?

**Comment by FirefoxMetzger** [comment]
@wjp I agree with @hmaarrfk . If you could give us access to such an image and permission to distribute it as part of our test suite then I can write a test for it so that we can be sure these images are handled correctly moving forward. (All TIFF writers I have access to refuse to write a 0 denominator.)

**Comment by wjp** [comment]
I'm attaching a 288kb sample tif for testing purposes, and I hereby give permission to distribute this should you want to do so.
[metadata_resolution_test.zip](https://github.com/imageio/imageio/files/8125439/metadata_resolution_test.zip)

**Comment by hmaarrfk** [comment]
I added a few tests that should cover our bases. Let me know if there is anything else to do here.

**Review by FirefoxMetzger — COMMENTED** [review]
Thanks for the bugfix @hmaarrfk !

I am pretty happy so far. There is one discussion point left, but if we manage to agree on a good way forward then we can get this merged before the next CD cycle (Monday).

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  the denominator of the metadata. Such images will not have a key
+  `"resolution"` in the metadata dictionary.
+  ([#753](https://github.com/imageio/imageio/pull/753)))
+
+### Feature
+* placeholder
+
+### Other
+* placeholder
+
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply CHANGELOG.md:23]: How do I control the content there? Is there a stub file?

**Comment by hmaarrfk** [comment]
@FirefoxMetzger thank you for bringing up the `policy` behavior. 

I've been trying to push on "metadata" for a while, but ultimately, I find that it is always a "secondary" interest to the data itself. Many people want the data first, metadata second. It seems strange to error by default on a metadata bug, often in external software you don't necessarily control because their interpretation of what is allowed isn't entirely correct.

It may be that our interpretation of allowed fields is wrong as the standard evolves.

For this reason, I would avoid "strict" and prefer to default on "warn".

**Comment by hmaarrfk** [comment]
I've also seen people use ipython magic `%capture` to silence loud warnings I have asking people to help in various edge cases.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (FirefoxMetzger) [reply CHANGELOG.md:23]: Via the commit message once the PR gets merged. 

We tag them using scipy tags (MAINT, BUG, FEAT/ENH, DEV, ...) and the CD will sort the item into the appropriate section. The bullet point is equal to the commit message, and CD will append a link to this PR where you find all the details if you need them.

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply CHANGELOG.md]: thx.

**Comment by FirefoxMetzger** [comment]
> I've been trying to push on "metadata" for a while, but ultimately, I find that it is always a "secondary" interest to the data itself. Many people want the data first, metadata second.

Not sure if you were aware, but we are making headway with metadata in ImageIO. The docs need final approval from @almarklein , but the functions are already merged: [`iio.improps`](https://imageio--718.org.readthedocs.build/en/718/reference/_core_v3/imageio.v3.improps.html#imageio.v3.improps) and [`iio.immeta`](https://imageio--718.org.readthedocs.build/en/718/reference/_core_v3/imageio.v3.immeta.html#imageio.v3.immeta) will serve you metadata and [`iio.imread`](https://imageio--718.org.readthedocs.build/en/718/reference/_core_v3/imageio.v3.imread.html#imageio.v3.imread) will simply return a plain old ndarray. 

> often in external software you don't necessarily control because their interpretation of what is allowed isn't entirely correct.

I agree, which is why we need a way of being not strict. I like being strict because it helps becoming aware of issues; even if you cant fix them, you can at least report them so that upstream knows that the problem exists. That said, I see how being strict can be annoying for the average user, so maybe `warn` is the better default. If you like the idea, would you like to implement the kwarg for the tiffile plugin, or should I?

Also, would it be an idea to set a resolution of (1, 0) to `np.nan` instead of dropping the field? (No idea if that is a more sensible choice, it really is just an open suggestion.)

**Comment by hmaarrfk** [comment]
> I agree, which is why we need a way of being not strict. I like being strict because it helps becoming aware of issues; even if you cant fix them, you can at least report them so that upstream knows that the problem exists. That said, I see how being strict can be annoying for the average user, so maybe warn is the better default. If you like the idea, would you like to implement the kwarg for the tiffile plugin, or should I?

You can push to my repo if you have time. I'm sure propagating the keyword argument will take me some time to figure out.

> Also, would it be an idea to set a resolution of (1, 0) to np.nan instead of dropping the field? (No idea if that is a more sensible choice, it really is just an open suggestion.)

I've really struggled with incorporating `np.nan` in "python" objects. The resolution is a `tuple` containing 2 python floats and a string. I feel that `np.nan` changes the data type. We can likely use `math.nan` but we can likely use `nan` and `inf` if you like:
```
In [5]: np.asarray(1) / np.asarray(0)
<ipython-input-5-058e4fa61a41>:1: RuntimeWarning: divide by zero encountered in true_divide
  np.asarray(1) / np.asarray(0)
inf

In [6]: np.asarray(0) / np.asarray(0)
<ipython-input-6-bdc5991fde46>:1: RuntimeWarning: invalid value encountered in true_divide
  np.asarray(0) / np.asarray(0)
nan
```
seems to be consistent with numpy conventions.

**Comment by FirefoxMetzger** [comment]
> You can push to my repo if you have time. I'm sure propagating the keyword argument will take me some time to figure out.

Actually, sleeping over this for a night, I think I have a better/faster idea: In this PR we hard-code a warning when we ignore `Resolution` (could you add it?). I will then go and create a new issue for the `policy` kwarg and develop it as a new feature in a separate PR.

In doing so, we can get this merged in today (or tomorrow in your timezone I guess), and I can pick up `policy` once I get around to porting the tifffile plugin. Thoughts?

---

> The resolution is a tuple containing 2 python floats and a string. I feel that np.nan changes the data type. 

That's a good point; we shouldn't change the dtype and use `float("nan")` and `float("inf")` in those cases ... Naa, lets leave it as is for now. If somebody expects to use a 0 denominator in their application and is surprised that we do not load it, we can always add it in later. Would be a small/easy change.

**Comment by FirefoxMetzger** [comment]
@hmaarrfk I've added a warning to your branch that will warn when resolution is being ignored. This is good enough for now (it solves my worries about silently ignoring metadata). Once CI is happy I will merge this in :)

I will also create a new issue regarding the `policy` kwarg for TIFF and other plugins and see what other people think. I will then work that in once I get around to porting our tifffile plugin.

**Pull request merged**